// @ts-rest/fastify/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var core = require('@ts-rest/core');
var zod = require('zod');

class RequestValidationError extends Error {
    constructor(pathParams, headers, query, body) {
        super('[ts-rest] request validation failed');
        this.pathParams = pathParams;
        this.headers = headers;
        this.query = query;
        this.body = body;
    }
}
const RequestValidationErrorSchema = zod.z.object({
    pathParameterErrors: core.ZodErrorSchema.nullable(),
    headerErrors: core.ZodErrorSchema.nullable(),
    queryParameterErrors: core.ZodErrorSchema.nullable(),
    bodyErrors: core.ZodErrorSchema.nullable(),
});
const isAppRouteImplementation = (obj) => {
    return typeof obj === 'function';
};
const validateRequest = (request, reply, schema, options) => {
    const paramsResult = core.checkZodSchema(request.params, schema.pathParams, {
        passThroughExtraKeys: true,
    });
    const headersResult = core.checkZodSchema(request.headers, schema.headers, {
        passThroughExtraKeys: true,
    });
    const queryResult = core.checkZodSchema(options.jsonQuery
        ? core.parseJsonQueryObject(request.query)
        : request.query, schema.query);
    const bodyResult = core.checkZodSchema(request.body, 'body' in schema ? schema.body : null);
    if (!paramsResult.success ||
        !headersResult.success ||
        !queryResult.success ||
        !bodyResult.success) {
        throw new RequestValidationError(paramsResult.success ? null : paramsResult.error, headersResult.success ? null : headersResult.error, queryResult.success ? null : queryResult.error, bodyResult.success ? null : bodyResult.error);
    }
    return {
        paramsResult,
        headersResult,
        queryResult,
        bodyResult,
    };
};
const RouterEmbeddedContract = Symbol('RouterEmbeddedContract');
const initServer = () => ({
    router: (contract, routes) => ({
        ...routes,
        [RouterEmbeddedContract]: contract,
    }),
    route: (route, implementation) => implementation,
    registerRouter: (contract, routerImpl, app, options = {
        logInitialization: true,
        jsonQuery: false,
        responseValidation: false,
        requestValidationErrorHandler: 'combined',
    }) => {
        const { hooks = {}, ...restOfOptions } = options;
        Object.entries(hooks).forEach(([hookName, hookOrHookArray]) => {
            if (Array.isArray(hookOrHookArray)) {
                hookOrHookArray.forEach((hook) => {
                    // @ts-expect-error - function expects specific hook names rather than just a string
                    app.addHook(hookName, hook);
                });
                return;
            }
            else {
                // @ts-expect-error - function expects specific hook names rather than just a string
                app.addHook(hookName, hookOrHookArray);
            }
        });
        recursivelyRegisterRouter(routerImpl, contract, [], app, restOfOptions);
        app.setErrorHandler(requestValidationErrorHandler(options.requestValidationErrorHandler));
    },
    plugin: (router) => (app, opts = {
        logInitialization: true,
        jsonQuery: false,
        responseValidation: false,
        requestValidationErrorHandler: 'combined',
    }, done) => {
        const embeddedContract = router[RouterEmbeddedContract];
        const { hooks = {}, ...restOfOptions } = opts;
        Object.entries(hooks).forEach(([hookName, hookOrHookArray]) => {
            if (Array.isArray(hookOrHookArray)) {
                hookOrHookArray.forEach((hook) => {
                    // @ts-expect-error - function expects specific hook names rather than just a string
                    app.addHook(hookName, hook);
                });
                return;
            }
            else {
                // @ts-expect-error - function expects specific hook names rather than just a string
                app.addHook(hookName, hookOrHookArray);
            }
        });
        recursivelyRegisterRouter(router, embeddedContract, [], app, restOfOptions);
        app.setErrorHandler(requestValidationErrorHandler(opts.requestValidationErrorHandler));
        done();
    },
});
const requestValidationErrorHandler = (handler = 'combined') => {
    return (err, request, reply) => {
        if (err instanceof RequestValidationError) {
            if (handler === 'combined') {
                return reply.status(400).send({
                    pathParameterErrors: err.pathParams,
                    headerErrors: err.headers,
                    queryParameterErrors: err.query,
                    bodyErrors: err.body,
                });
            }
            else {
                return handler(err, request, reply);
            }
        }
        else {
            return reply.send(err);
        }
    };
};
/**
 * @param routeImpl - User's implementation of the route
 * @param appRoute - the `ts-rest` contract for this route (e.g. with Path, params, query, body, etc.)
 * @param app - the fastify instance to register the route on
 * @param options - options for the routers
 */
const registerRoute = (routeImpl, appRoute, app, options) => {
    if (options.logInitialization) {
        app.log.info(`[ts-rest] Initialized ${appRoute.method} ${appRoute.path}`);
    }
    const handler = isAppRouteImplementation(routeImpl)
        ? routeImpl
        : routeImpl.handler;
    const hooks = isAppRouteImplementation(routeImpl)
        ? {}
        : routeImpl.hooks || {};
    const route = {
        ...hooks,
        method: appRoute.method,
        url: appRoute.path,
        config: {
            tsRestRoute: appRoute,
        },
        handler: async (request, reply) => {
            const validationResults = validateRequest(request, reply, appRoute, options);
            let result;
            try {
                result = await handler({
                    // eslint-disable-next-line @typescript-eslint/no-explicit-any
                    params: validationResults.paramsResult.data,
                    query: validationResults.queryResult.data,
                    // eslint-disable-next-line @typescript-eslint/no-explicit-any
                    headers: validationResults.headersResult.data,
                    request,
                    // eslint-disable-next-line @typescript-eslint/no-explicit-any
                    body: validationResults.bodyResult.data,
                    reply,
                    appRoute,
                });
            }
            catch (e) {
                if (e instanceof core.TsRestResponseError) {
                    result = {
                        status: e.statusCode,
                        body: e.body,
                    };
                }
                else {
                    throw e;
                }
            }
            const statusCode = result.status;
            let validatedResponseBody = result.body;
            if (options.responseValidation) {
                const response = core.validateResponse({
                    appRoute,
                    response: {
                        status: statusCode,
                        body: result.body,
                    },
                });
                validatedResponseBody = response.body;
            }
            const responseType = appRoute.responses[statusCode];
            if (core.isAppRouteNoBody(responseType)) {
                return reply.status(statusCode).send();
            }
            if (core.isAppRouteOtherResponse(responseType)) {
                reply.header('content-type', responseType.contentType);
            }
            return reply.status(statusCode).send(validatedResponseBody);
        },
    };
    app.route(route);
};
/**
 *
 * @param routerImpl - the user's implementation of the router
 * @param appRouter - the `ts-rest` contract for this router
 * @param path - the path to the current router, e.g. ["posts", "getPosts"]
 * @param fastify  - the fastify instance to register the route on
 * @param options
 */
const recursivelyRegisterRouter = (routerImpl, appRouter, path, fastify, options) => {
    if (typeof routerImpl === 'object' &&
        typeof (routerImpl === null || routerImpl === void 0 ? void 0 : routerImpl['handler']) !== 'function') {
        for (const key in routerImpl) {
            recursivelyRegisterRouter(routerImpl[key], appRouter[key], [...path, key], fastify, options);
        }
    }
    else if (typeof routerImpl === 'function' ||
        typeof (routerImpl === null || routerImpl === void 0 ? void 0 : routerImpl['handler']) === 'function') {
        registerRoute(routerImpl, appRouter, fastify, options);
    }
};

exports.RequestValidationError = RequestValidationError;
exports.RequestValidationErrorSchema = RequestValidationErrorSchema;
exports.initServer = initServer;
